### PR TITLE
Refactor main.go with  oklog/pkg/group actors pattern

### PR DESCRIPTION
**Pull request by krasi-georgiev**
At the moment main.go and some internal packages use a  mix of different concurrency patterns and with this change I am hoping to make it more structured, easier to follow and enforce any further Run methods to follow the same pattern.

a brief talk with examples for this pattern
https://youtu.be/LHe1Cb_Ud_M?t=1487

TODO:
 - [ ] ~~refactor some Run-ers to remove the need for an additional blocking channel~~ (will be another PR)
 - [x] shutdown the query engine before the rule manager as per this comment
- [x] is the execution order correct ? 
```
// Shutting down the query engine before the rule manager will cause pending queries
// to be canceled and ensures a quick shutdown of the rule manager.
```

 - [ ] ~~add some tests ?~~

### CODE REVIEW

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -275,104 +271,182 @@ func main() {
...
+				select {
+				case <-term:
+					level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
+				case <-waitShutdown:
+					break
+				}
+				return nil
+			},
+			func(err error) {
+			},
```
Comment (peterbourgon) [inline cmd/prometheus/main.go]: It's invalid to have an empty interrupt function.

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -275,104 +271,182 @@ func main() {
...
+					level.Error(logger).Log("msg", "Opening storage failed", "err", err)
+				} else {
+					localStorage.Set(db)
+				}
+				close(dbOpen)
+				<-waitShutdown
+				return nil
+			},
+			func(err error) {
+				if err := fanoutStorage.Close(); err != nil {
```
Comment (peterbourgon) [inline cmd/prometheus/main.go]: How does this terminate the execute function, above?

File: cmd/prometheus/main.go [hunk 3]
```diff
@@ -275,104 +271,182 @@ func main() {
...
+				return nil
+			},
+			func(err error) {
+				notifier.Stop()
+			},
+		)
+		g.Add(
+			func() error {
+				ruleManager.Run()
+				<-waitShutdown
```
Comment (peterbourgon) [inline cmd/prometheus/main.go]: You appear to have several actors waiting to return by receiving on the same chan. This is invalid. Each actor must have a lifecycle independent of the others.

Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: Oh boy , different channel for each actor...
Is this because if the actors exit at the same time the interrupt function might not run ?

Comment (peterbourgon) [reply cmd/prometheus/main.go]: No. One actor is guaranteed to exit first. When that happens, the interrupt functions for all actors will be invoked sequentially.

Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: in that case why do I need a separate channel for each actor?

I will refactor some actors to be blocking and remove the `<-waitShutdown` , but others wouldn't make sence to be blocking so wanted to use a single channel for simplicity.

Some of the code in the blocks is not even actors and just some bootstrapping , but wanted to put all in groups so it is easier to follow and share the same structure (lexical scoping)

Comment (peterbourgon) [reply cmd/prometheus/main.go]: The semantics of the Add method's parameters [are documented here](https://godoc.org/github.com/oklog/oklog/pkg/group).

> Add an actor (function) to the group. Each actor must be pre-emptable by an interrupt function. That is, [b]if interrupt is invoked, execute should return[/b]. Also, it must be safe to call interrupt even after execute has returned.

If your execute function is blocking in a chan recv on a single, shared waitForShutdown chan, then the bolded requirement is not met.

Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: Yes I understand the semantics,but why is this important in this implementation?  Maybe give an example worst case of what will happen if this condition is not met.

`<-waitShutdown` is there just as an ugly hack and doesn't matter if the interruptor is run after or before the executor has exited.

Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: or maybe give a suggestion how to improve the implementation without any major refactoring of the executors to begin with.

Comment (peterbourgon) [reply cmd/prometheus/main.go]: You can envision the internals of the group as something like this:

1. Launch all N actor execute funcs in their own goroutines
1. Wait to receive first error from an execute func
1. For each actor
    1. Invoke interrupt func
    1. Receive error from execute func

The fact that the current version of the group code invokes all interrupt funcs first, and then waits to receive (N-1) errors second, is kind of an implementation detail, done to avoid adding per-actor tracking scaffolding. The implementation could easily change in the future. It's not safe to violate the bolded requirement; that's why it's explicitly documented as a requirement.

Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: got , it!

my confusion came form the current `group` implementation

this comment is a good clue about the design goals.
> // The error is passed to the interrupt functions

**Comment by krasi-georgiev** [comment]
@peterbourgon grouped some code and added separate channel for each interruptor. thanks for the review

**Comment by krasi-georgiev** [comment]
cc @fabxc , @brian-brazil 
trying to make sense of this comment
> The notifier is a dependency of the rule manager. 
> It has to be started before and torn down afterwards.

https://github.com/prometheus/prometheus/blob/dev-2.0/cmd/prometheus/main.go#L334-L340
if both are run in a go routines how is it guaranteed that 'notifier.Run()' will run before 'ruleManager.Run()' ?

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -282,104 +278,182 @@ func main() {
...
+	prometheus.MustRegister(configSuccess)
+	prometheus.MustRegister(configSuccessTime)
 
-		db, err := tsdb.Open(
-			cfg.localStoragePath,
-			log.With(logger, "component", "tsdb"),
-			prometheus.DefaultRegisterer,
-			&cfg.tsdb,
+	var g group.Group
+	{
```
Comment (grobie) [inline cmd/prometheus/main.go]: I believe the idea was to use one lexical scope per actor, not one for all actors. This would enable you to use the same variable name for interrupt channels instead of the current s1, s2, ... pattern. Channels like `hupReady` which are used in multiple actors would need to be defined in the scope of the main function.

File: cmd/prometheus/main.go [hunk 5]
```diff
@@ -282,104 +278,182 @@ func main() {
...
-	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
-	select {
-	case <-term:
-		level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
-	case <-webHandler.Quit():
-		level.Warn(logger).Log("msg", "Received termination request via web service, exiting gracefully...")
-	case err := <-errc:
-		level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+	if err := g.Run(); err != nil {
+		level.Error(logger).Log("msg", "Group interrupt", "err", err)
```
Comment (grobie) [inline cmd/prometheus/main.go]: As a Prometheus user, without knowing the code and specific implementation here, I would not know what "group" is referring to.

File: cmd/prometheus/main.go [hunk 6]
```diff
@@ -439,8 +508,8 @@ func computeExternalURL(u, listenAddr string) (*url.URL, error) {
 		u = fmt.Sprintf("http://%s:%s/", hostname, port)
 	}
 
-	if startsOrEndsWithQuote(u) {
-		return nil, fmt.Errorf("URL must not begin or end with quotes")
+	if ok := govalidator.IsURL(u); !ok {
```
Comment (grobie) [inline cmd/prometheus/main.go]: What is the reason to revert #3199 and bring back govalidator? This seems to be outside of the scope of this PR.

File: cmd/prometheus/main.go [hunk 7]
```diff
@@ -282,104 +278,182 @@ func main() {
...
+			func() error {
+				db, err := tsdb.Open(
+					cfg.localStoragePath,
+					log.With(logger, "component", "tsdb"),
+					prometheus.DefaultRegisterer,
+					&cfg.tsdb,
+				)
+				if err != nil {
+					level.Error(logger).Log("msg", "Opening storage failed", "err", err)
+					return nil
```
Comment (grobie) [inline cmd/prometheus/main.go]: Why is nil being returned here and not the error?

File: cmd/prometheus/main.go [hunk 8]
```diff
@@ -282,104 +278,182 @@ func main() {
...
+				targetManager.Stop()
+				close(s6)
+			},
+		)
+		g.Add(
+			func() error {
+				if err := webHandler.Run(ctx); err != nil {
+					level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+				}
+				return nil
```
Comment (grobie) [inline cmd/prometheus/main.go]: Why is nil being returned here and not the error?

File: cmd/prometheus/main.go (hunk 4 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: the original implementation was with a single channel, but if you read peterbourgon's original comments few commits behind you will see why we need a separate channel as a interruptor.

File: cmd/prometheus/main.go (hunk 8 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: I tried to return and print the error but you loose the extra `msg` string
`"msg", "Error starting web server, exiting gracefully",`

File: cmd/prometheus/main.go (hunk 5 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: you mean this line `level.Error(logger).Log("msg", "Group interrupt", "err", err)` ? 
this is not even used as you can see that all groups return `nil` for the error. 

I am still not sure how to organise the group errors as if we move the errors to be handled by the group package we would loose the extra `"msg"` field.

Open to suggestions.

File: cmd/prometheus/main.go (hunk 4 above)
Comment (grobie) [reply cmd/prometheus/main.go]: You'd still have separate channels. He talks about the advantage of using individual lexical groups in his talk.

```go
{
    cancel := make(chan struct{})
    g.Add(
        // ...
    )
}
{
    cancel := make(chan struct{})
    g.Add(
        // ...
    )
}
// ... other groups
```

File: cmd/prometheus/main.go (hunk 6 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: not sure what happened will fix it.

File: cmd/prometheus/main.go (hunk 8 above)
Comment (grobie) [reply cmd/prometheus/main.go]: Exiting gracefully is expected, I don't think that needs to be logged. You could just wrap the error message `return fmt.Errorf("Error starting web server: %s", err)`.

File: cmd/prometheus/main.go (hunk 4 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: changed and it looks better like this.

File: cmd/prometheus/main.go [hunk 9]
```diff
@@ -282,104 +277,193 @@ func main() {
...
-			}
-		}
-	}()
+	prometheus.MustRegister(configSuccess)
+	prometheus.MustRegister(configSuccessTime)
 
 	// Start all components while we wait for TSDB to open but only load
 	// initial config and mark ourselves as ready after it completed.
 	dbOpen := make(chan struct{})
+	// Wait for reload or termination signals. Start the handler for SIGHUP as
```
Comment (grobie) [inline cmd/prometheus/main.go]: This channel/handler doesn't seem to be related to termination signals, but only to reloads.

File: cmd/prometheus/main.go [hunk 10]
```diff
@@ -282,104 +277,193 @@ func main() {
...
+	// early as possible, but ignore it until we are ready to handle reloading
+	// our config.
+	hupReady := make(chan bool)
```
Comment (grobie) [inline cmd/prometheus/main.go]: Nit: Reloads can be triggered either via SIGHUP or a web request. A name like `reloadReady` seems to be a better name here.

Comment (grobie) [reply cmd/prometheus/main.go]: Nit: We use `chan struct{}` everywhere else.

File: cmd/prometheus/main.go [hunk 11]
```diff
@@ -282,104 +277,193 @@ func main() {
...
-	if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
-		level.Error(logger).Log("msg", "Error loading config", "err", err)
-		os.Exit(1)
+			},
+			func(err error) {
+				close(cancel)
+			},
+		)
 	}
+	{
```
Comment (grobie) [inline cmd/prometheus/main.go]: Nit: The usage of empty lines is inconsistent across this code. I'd vote to not add empty lines after `{` consistently.

File: cmd/prometheus/main.go [hunk 12]
```diff
@@ -282,104 +277,193 @@ func main() {
...
-	term := make(chan os.Signal, 1)
-	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
-	select {
-	case <-term:
-		level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
-	case <-webHandler.Quit():
-		level.Warn(logger).Log("msg", "Received termination request via web service, exiting gracefully...")
-	case err := <-errc:
-		level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+				// Wait for reload or termination signals.
```
Comment (grobie) [inline cmd/prometheus/main.go]: Nit: There are two comments for a single line of code which is already pretty readable in itself. I'd vote to replace this line here with `// Unblock reload handler.` and remove the comment below.

File: cmd/prometheus/main.go [hunk 13]
```diff
@@ -282,104 +277,193 @@ func main() {
...
+				if err := fanoutStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
+				}
+				close(cancel)
+			},
+		)
+	}
+	{
+		// The notifier is a dependency of the rule manager. It has to be
+		// started before and torn down afterwards.
```
Comment (grobie) [inline cmd/prometheus/main.go]: What is the purpose of this comment? I can't see how this requirement is enforced at the moment.

File: cmd/prometheus/main.go [hunk 14]
```diff
@@ -282,104 +277,193 @@ func main() {
...
+				return nil
+			},
+			func(err error) {
+				ruleManager.Stop()
+				close(cancel)
+			},
+		)
 	}
+	{
+		//TODO(krasi) refactor targetManager.Run() to be blocking to avoid using an extra blocking channel
```
Comment (grobie) [inline cmd/prometheus/main.go]: Nit: The Prometheus project has been pretty pedantic about following the golang comment style so far. Please leave a space after `//` and end all comments on a dot.

Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: ok , I will try to make `targetManager.Run()` blocking so I would remove the comment anyway.

File: cmd/prometheus/main.go (hunk 13 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: this is a copy and paste form the original implementation and I left it so that I don't forget to ask the same question someone with more context  :) . It is in the TODO list in the PR comment

https://github.com/prometheus/prometheus/blob/dev-2.0/cmd/prometheus/main.go#L334-L340
if both are run in a go routines how is it guaranteed that notifier.Run() will run before ruleManager.Run() ?

comment added with this commit
#7e1b39c

**Comment by krasi-georgiev** [comment]
> All the log lines like "Starting TSDB" (maybe use "time series database" instead) are pretty important to understand what's going on during startup/end of the program cycle. Please add back all the "Starting ..." and "Started ..." log lines. The actor functions seem to be a good place to do that consistently.

These were  a bit inconsistent, some are missing the  `Starting`  log and others are missing the  `Stoping` log



It is usefull to know which TSDB backend is  used , but why do we need the start up and stopping messages ? Aren't these needed just for debugging (maybe only when debug is enabled) ?

with the concurrency groups the logs were a bit messed up and difficult to follow
```
Stopping Target Manager
Stopping Rule Manager
Stoped Rule Manager
Stopped Target Manager
```

**Comment by grobie** [comment]
Given target and rule manager load quasi instantaneously, and as I can't
remember a bug where loading these actors was delayed, I'd be fine leaving
these out. I feel strongly about having start and started messages for the
TSDB. Without these, it'd be unclear what's going on and what the state of
the server is.

**Comment by krasi-georgiev** [comment]
@grobie 
wouldn't be better to show all starting, stopping logs only with debug enable?

otherwise I think the error messages are sufficient to show that something is not OK.
`level.Error(logger).Log("msg", "Opening storage failed", "err", err)`

I always find less verbosity by default to be a better approach. :)

**Comment by grobie** [comment]
I agree with your general sentiment on log levels, but there is a difference between verbosity and lack of user feedback. Starting the TSDB can take several minutes, it'd be completely unclear what the server is doing and whether it's ready already without the log lines. I'll object merging this PR without bringing back the start log lines for at least the TSDB actor.

File: cmd/prometheus/main.go (hunk 13 above)
Comment (grobie) [reply cmd/prometheus/main.go]: IIRC the important part of that commit was not to fix the start order (the order doesn't seem to be deterministic at the moment), but the shutdown order. It's important to first stop the rule manager before the notification handler, as the first will panic if the latter isn't running (anymore). Due to using defer and blocking close methods, that order is ensured.

To not add a regression and bring back the possibility of panics during shutdown, something needs to be done here. Either synchronizing the too shutdown functions by introducing another channel (similar to reloadReady), or by redesigning the implementation of the two actors. I think the first makes more sense in the scope of this PR.

**Comment by krasi-georgiev** [comment]
@grobie sure no problem I added the tsdb logs , you have more context here so experience always beats preferences, 
It would be interesting to hear the other maintainers as well and I will add the rest of the logs if needed.

**Comment by brian-brazil** [comment]
I agree with @grobie, these log lines should remain as they help debug startup issues.

**Comment by fabxc** [comment]
Yup, we had plenty of regressions where some components deadlocked on startup etc. and the log lines clearly indicated which one was the offender.
Just enabling it for debug logs makes it harder for us to figure out reported issues, because we need another feedback loop when asking people to retry with debug level enabled.

**Comment by krasi-georgiev** [comment]
ok  great I added them again

I would like at some point to refactor the  async Run methods to be blocking to avoid using an extra channel but that should be another PR when I build up some context on the internal executions.

but it would be great if you could comment on some of the other items in the TODO list

File: cmd/prometheus/main.go (hunk 13 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: somehow I missed this comment and was waiting for some feedback...
anyway , the good news is that  the interrupts are blocking and run in a sequence so I just reordered the blocks and now it is guaranteed that `notifier.Stop()` is called only after `ruleManager.Stop()` exited.

here is the Actors interrupt implementation
https://github.com/oklog/oklog/blob/v0.2.2/pkg/group/group.go#L46-L48

Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: feel free to change any comments format/text if something doesn't sound right.

**Comment by krasi-georgiev** [comment]
do you think we need some additional tests with this change ?

File: cmd/prometheus/main.go [hunk 15]
```diff
@@ -282,104 +277,197 @@ func main() {
...
+			},
+			func(err error) {
+				ruleManager.Stop()
+				close(cancel)
+			},
+		)
+	}
+	{
+		// Keep this interrupt after the ruleManager.Stop().
+		// calling notifier.Stop() before ruleManager.Stop() will cause a panic if the ruleManager isn't running (anymore).
```
Comment (grobie) [inline cmd/prometheus/main.go]: @krasi-georgiev That's the only thing left, right?
